### ISSUE TITLE
J4TS related issue

### ISSUE THREAD

**Issue by isakovarseniy** [cincheo/jsweet#391]
Hi 
I got following exception

candies/typings/META-INF/resources/typings/j4ts/0.6.0-SNAPSHOT/bundle.d.ts(3141,13): error TS2451: Cannot redeclare block-scoped variable 'Map'.
../../../../../../servers/node-v6.11.3-darwin-x64/lib/node_modules/typescript/lib/lib.es6.d.ts(4621,11): error TS2451: Cannot redeclare block-scoped variable 'Map'.

```
{
  "compilerOptions": {
    "baseUrl": "./",
    "module" : "es2015",
    "target": "ES6",
    "moduleResolution": "classic",
    "sourceMap": true,
    "declaration": true
  },
  "include": [
    "src/**/*.ts",
    "candies/typings/META-INF/resources/typings/j4ts/0.6.0-SNAPSHOT/bundle.d.ts"


  ],
  "exclude": [
    "node_modules"
  ]
}
```

Thanks
Arseniy Isakov

**Comment by renaudpawlak** [comment]
I don't have enough details here to understand the issue. I have to look at what you did for packaging J4TS with modules, but it looks like a conflict between the Java and JavaScript maps. I am surprised that you need to include ``bundle.d.ts``, since this file is not generated when using modules...

**Comment by isakovarseniy** [comment]
Actually I mixed ES6 module with j4ts bundle.  Because j4ts is not a ES6

```
export class InsertQuery extends Query<InsertQuery> {
    /*private*/ columnsValues : java.util.LinkedHashMap<string, any> = <any>(<java.util.LinkedHashMap<string, any>>new java.util.LinkedHashMap<string, any>());

.......
```

**Comment by isakovarseniy** [comment]
in J4TS you have following class
```
package javaemul.internal.globals;
import jsweet.lang.Ambient;

public class Globals {

	@Ambient
	public static def.js.Object Map;
	
}
```

As a result its create variable Map
line 3040 in candies/typings/META-INF/resources/typings/j4ts/0.5.0-SNAPSHOT/bundle.d.ts
```ts
declare let Map: Object;

```

**Comment by renaudpawlak** [comment]
Actually the problem is that I use the Map object in the implementation of maps. Unfortunately in typescript there is no conditional declarations. So I think that the Map declaration should be removed and that the Map object should be accessed in a un typed way. I can deal with it if you want.

**Comment by isakovarseniy** [comment]
Sure. Thanks.

**Comment by renaudpawlak** [comment]
Umm. Maybe it will compile but it probably will not work at runtime because this variable tries to bind to Map, which will not work if you change the name...

**Comment by renaudpawlak** [comment]
Ok I have just pushed a fix in J4TS. I don't have time to test it (Maps)... so I will leave it up to map users or the appropriate unit tests. 

By the way, there is a problem with the unit tests because they don't compile under Eclipse. I am not sure what it is but we should solve this if we want the project to evolve smoothly.... Any clue why it does not work and how can we extend the unit tests to more APIs?

Please tell me if there is an issue with my fix.

**Comment by isakovarseniy** [comment]
I am not sure about eclipse. It shows me compilation error.

**Comment by renaudpawlak** [comment]
Your ``getJsMapConstructor`` implementation is the same idea as mine. If mine does not work for some reason, please don't hesitate to replace. I don't see why you need to wrap in a function though (but it does not matter).

**Comment by renaudpawlak** [comment]
Actually, about Eclipse, the problem is in the ``BaseJreTest``, mainly in the ``allTests`` method, because it uses the reflect and stream API in a way that is not supported by J4TS (which overrides some default Java API). So the ``Map.forEach`` and ``Class.getDeclaredMethods`` methods are not found. 

Besides there were some minor typing issues in the stream API that raised a compilation error in the tests. I have fixed it in https://github.com/j4ts/j4ts/commit/80ada99720ac45750d88f2676c3f04fa9426b017.

**Comment by renaudpawlak** [comment]
So, all that being said, I don't understand if you are transpiting the tests in JSweet and execute them in JavaScript, or if you are running them in Java -- which is not really useful for our case.
It is clearly a good idea to setup JUnit tests for J4TS, but we need a way to transpile them and run them in JavaScript. What's your status on this?

BTW, you may want to look at this work: https://github.com/renaudpawlak/tura/blob/4.3.0/platform/platform-ts/test/test-ts-sql-dsl/jsweet_extension/org/tura/jsweet/JUnitAdapter.java. It is a JSweet extension to support JUnit.

**Comment by renaudpawlak** [comment]
Oops, my mistake. Of course you know about it, because, you did it :) It's early in the morning here and I got confused ;) So, I guess that were I get confused is that I don't see your JUnit extension in J4TS so, maybe you are compiling the tests using an external project?

**Comment by isakovarseniy** [comment]
Yes I did it :). First of all you implementation of getJsMapConstructor() gives me an error during execution " window variable is not defined." It is working in implementation. I wrapped up in function to return value for eval. I am not very good with javascript. Probably there is a better way of doing this things. But this method working for sure,

**Comment by isakovarseniy** [comment]
I am using mocha for testing. You can see it in my tura project. There is a conflict between J4TS and mocha. I had to refactor J4TS (  I moved test.Test  -> j4ts.test.Test package). It is not committed yet to my project.

**Comment by thw0rted** [comment]
Is this supposed to be fixed?  I'm running jsweet 2.0.0 (based on the quickstart), and added a dep on j4ts 0.5.0. Building with `<targetVersion>ES5</targetVersion>` works but `<targetVersion>ES6</targetVersion>` gives me 

````
2018-04-18 16:05:37.037 INFO  JSweetTranspiler:1403 - ../../.jsweet/candies/typings/META-INF/resources/typings/j4ts/0.5.0/bundle.d.ts(3040,13): error TS2451: Cannot redeclare block-scoped variable 'Map'.
2018-04-18 16:05:37.037 ERROR output:55 - cannot redeclare block-scoped variable 'Map' at ..\..\.jsweet\candies\typings\META-INF\resources\typings\j4ts\0.5.0\bundle.d.ts(3040)
2018-04-18 16:05:37.037 INFO  JSweetTranspiler:1403 - ../../../../Users/James/.jsweet-node_modules/node_modules/typescript/lib/lib.es6.d.ts(4724,11): error TS2451: Cannot redeclare block-scoped variable 'Map'.
2018-04-18 16:05:37.037 ERROR output:55 - cannot redeclare block-scoped variable 'Map' at ..\..\..\..\Users\James\.jsweet-node_modules\node_modules\typescript\lib\lib.es6.d.ts(4724)
2018-04-18 16:05:37.037 INFO  JSweetTranspiler:1403 - ../../../../Users/James/.jsweet-node_modules/node_modules/typescript/lib/lib.es6.d.ts(4739,13): error TS2451: Cannot redeclare block-scoped variable 'Map'.
2018-04-18 16:05:37.037 ERROR output:55 - cannot redeclare block-scoped variable 'Map' at ..\..\..\..\Users\James\.jsweet-node_modules\node_modules\typescript\lib\lib.es6.d.ts(4739)
2018-04-18 16:05:37.037 INFO  JSweetTranspiler:1403 - ../../../../Users/James/.jsweet-node_modules/node_modules/typescript/lib/lib.es6.d.ts(4902,11): error TS2451: Cannot redeclare block-scoped variable 'Map'.
2018-04-18 16:05:37.037 ERROR output:55 - cannot redeclare block-scoped variable 'Map' at ..\..\..\..\Users\James\.jsweet-node_modules\node_modules\typescript\lib\lib.es6.d.ts(4902)
2018-04-18 16:05:37.037 INFO  JSweetTranspiler:1403 - ../../../../Users/James/.jsweet-node_modules/node_modules/typescript/lib/lib.es6.d.ts(5698,11): error TS2451: Cannot redeclare block-scoped variable 'Map'.
2018-04-18 16:05:37.037 ERROR output:55 - cannot redeclare block-scoped variable 'Map' at ..\..\..\..\Users\James\.jsweet-node_modules\node_modules\typescript\lib\lib.es6.d.ts(5698)
````


I don't think there's anything in my code causing this, maybe it's a fix that's still in SNAPSHOT but hasn't made it to a release yet?
